### PR TITLE
IPCM: console improvements

### PR DESCRIPTION
**Pull request by jmuchemb**
The 3 commits are atomic, in that we could only merge the first, or the first 2.
I implemented the 3rd one mainly for security reasons. Then, I'd have to update the wiki to use "socat" instead of "telnet".

### CODE REVIEW

**Comment by vmaffione** [comment]
Thanks, switching to UNIX sockets was needed indeed.

Then only concern I have is that people testing right tnow may don't want to change their script to use socat instead of telnet. A good strategy could be to let testers use pristine-1.4 and let's go ahead with this with pristine-1.5. What do the other guys think? @edugrasa  @dstaesse @kewinrausch, @vicalro, ... ?

General comment: please don't change the codestyle from
````
if (cond) {
  do;
}
````
to
````
if (cond)
  do;
````
since we decided it's safer to always have the curly brances, and also the codestyle change masks the real change.

File: rinad/src/ipcm/addons/console.cc [hunk 1]
```diff
@@ -166,57 +168,56 @@ IPCMConsole::~IPCMConsole() throw()
...
 	try {
-		sfd = socket(AF_INET, SOCK_STREAM, 0);
-		if (sfd < 0) {
-			ss  << " Error [" << errno <<
-				"] calling socket() " << endl;
-			throw rina::Exception();
-		}
+		sfd = socket(sa_un.sun_family, SOCK_STREAM, 0);
+		if (sfd < 0)
+			throw rina::Exception("socket");
```
Comment (vmaffione) [inline rinad/src/ipcm/addons/console.cc]: Could you please reuse the previous code to keep printing errno?

File: rinad/src/ipcm/addons/console.cc [hunk 2]
```diff
@@ -166,57 +168,56 @@ IPCMConsole::~IPCMConsole() throw()
...
-		ret = setsockopt(sfd, SOL_SOCKET, SO_REUSEADDR,
-				&optval, sizeof(optval));
-		if (ret < 0) {
-			ss  << " Error [" << errno <<
-				"] calling setsockopt(SOL_SOCKET, "
-				"SO_REUSEADDR, 1)" << endl;
-			FLUSH_LOG(WARN, ss);
-		}
+		if (bind(sfd, reinterpret_cast<struct sockaddr *>(&sa_un), len))
+			throw rina::Exception("bind");
```
Comment (vmaffione) [inline rinad/src/ipcm/addons/console.cc]: Same as above

File: rinad/src/ipcm/addons/console.cc [hunk 3]
```diff
@@ -166,57 +168,56 @@ IPCMConsole::~IPCMConsole() throw()
...
 
-		ret = bind(sfd, reinterpret_cast<struct sockaddr *>(&server_address),
-				sizeof(server_address));
-		if (ret < 0) {
-			ss  << " Error [" << errno <<
-				"] calling bind() " << endl;
-			throw rina::Exception();
-		}
+		if (listen(sfd, 5))
+			throw rina::Exception("listen");
```
Comment (vmaffione) [inline rinad/src/ipcm/addons/console.cc]: same as above

**Comment by kewinrausch** [comment]
I see no problem in having this merged in pristine-1.5 branch, as long as I can continue to use shell script to automate the testing phase. 

As long as the new method allows to script input into the console then I'm not against the change.

**Comment by vmaffione** [comment]
Also, an additional commit is required to change all the sample configuration files stored in tests/conf, in order to replace consolePort with consoleSocket.

**Comment by edugrasa** [comment]
Agree to merge in 1.5!

**Comment by vmaffione** [comment]
Also, an additional commit is required to change all the sample configuration files stored in tests/conf, in order to replace consolePort with consoleSocket.

**Comment by vmaffione** [comment]
Moreover, I see that runstatedir does not do the expected thing ("runstatedir" is resolved to "/" even with a non-empty $PREFIX).

The following patch to this PR works to me:
````
diff --git a/rinad/etc/Makefile.am b/rinad/etc/Makefile.am
index 4aa33a5..9382e2d 100644
--- a/rinad/etc/Makefile.am
+++ b/rinad/etc/Makefile.am
@@ -12,12 +12,10 @@ edit = $(SED)                                               \
         -e 's|@pkgdatadir[@]|$(pkgdatadir)|g'          \
         -e 's|@sysconfdir[@]|$(sysconfdir)|g'          \
         -e 's|@localstatedir[@]|$(localstatedir)|g'    \
-        -e 's|@runstatedir[@]|$(runstatedir)|g'                \
         -e 's|@libdir[@]|$(libdir)|g'                  \
         -e 's|@bindir[@]|$(bindir)|g'                  \
         -e 's|@sbindir[@]|$(bindir)|g'                 \
        -e 's|@prefix[@]|$(prefix)|g'                   \
-       -e 's|@localstatedir[@]|$(localstatedir)|g'     \
         -e 's|@LIBRINA_VERSION[@]|$(LIBRINA_VERSION)|g'
 
 CLEANFILES    =
diff --git a/rinad/etc/ipcmanager.conf.in b/rinad/etc/ipcmanager.conf.in
index 7e8c86f..55b3bd5 100644
--- a/rinad/etc/ipcmanager.conf.in
+++ b/rinad/etc/ipcmanager.conf.in
@@ -4,7 +4,7 @@
     "installationPath" : "@prefix@/bin",
     "libraryPath" : "@libdir@",
     "logPath" : "@localstatedir@/log",
-    "consoleSocket" : "@runstatedir@/ipcm-console.sock",
+    "consoleSocket" : "@localstatedir@/ipcm-console.sock",
     "pluginsPaths" : ["@libdir@/rinad/ipcp"]
   },
   "ipcProcessesToCreate" : [ {
diff --git a/rinad/src/ipcm/Makefile.am b/rinad/src/ipcm/Makefile.am
index 680a0be..4f363d9 100644
--- a/rinad/src/ipcm/Makefile.am
+++ b/rinad/src/ipcm/Makefile.am
@@ -14,7 +14,7 @@ CPPFLAGS_EXTRA =                                      \
        -DDEFAULT_BINDIR=\"$(bindir)\"                  \
        -DDEFAULT_LIBDIR=\"$(libdir)\"                  \
        -DDEFAULT_LOGDIR=\"$(localstatedir)/log\"       \
-       -DDEFAULT_RUNDIR=\"$(runstatedir)\"
+       -DDEFAULT_RUNDIR=\"$(localstatedir)/var\"
 
 #
 #IPCM
````
where I just reuse localstatedir.

File: rinad/src/ipcm/addons/console.cc (hunk 1 above)
Comment (jmuchemb) [inline rinad/src/ipcm/addons/console.cc]: The "catch" clause does log the errno value.

Comment (vmaffione) [inline rinad/src/ipcm/addons/console.cc]: Right! Sorry, I missed that.

**Comment by jmuchemb** [comment]
> we decided it's safer to always have the curly braces

Ok, project choice. I'll amend.

> Also, an additional commit is required to change all the sample configuration files stored in tests/conf, in order to replace consolePort with consoleSocket.

Oops, I missed that. I'll amend also, to keep the change atomic and because I already have other changes to do.

> Moreover, I see that runstatedir does not do the expected thing ("runstatedir" is resolved to "/" even with a non-empty $PREFIX).

Strange, it produces /var/run in my case. I'll change to use your code.

**Comment by vmaffione** [comment]
2016-03-31 16:37 GMT+02:00 Julien Muchembled <notifications@github.com>:

> we decided it's safer to always have the curly braces
>
> Ok, project choice. I'll amend.
>
> Also, an additional commit is required to change all the sample
> configuration files stored in tests/conf, in order to replace consolePort
> with consoleSocket.
>
> Oops, I missed that. I'll amend also, to keep the change atomic and
> because I already have other changes to do.
>
Sure, no worries.

> Moreover, I see that runstatedir does not do the expected thing
> ("runstatedir" is resolved to "/" even with a non-empty $PREFIX).
>
> Strange, it produces /var/run in my case. I'll change to use your code.
>
We can use the runstatedir approach. What I'm reporting is that for some
reason it fails to get the --prefix (that most of us use when configuring
IRATI). So maybe you could just rework your changes adding @prefix@ where
needed.


> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/IRATI/stack/pull/883#issuecomment-203968220>
>



-- 
Vincenzo Maffione

**Comment by vicalro** [comment]
Could you post an example on how to use socat to connect to the socket file?

**Comment by jmuchemb** [comment]
About runstatedir, localstatedir does not work for a system setup, because it would produce `/var/ipcm-console.sock`
So I looked at why runstatedir was empty for you and found that it's a recent feature (Autoconf 2.70 is not even released yet; it was backported by Debian in version 2.69-9).
I'd rather add the following lines just before the sed command:

    # --runstatedir is new to Autoconf 2.70
    runstatedir ?= $(localstatedir)/run

> Also, an additional commit is required to change all the sample configuration files stored in tests/conf, in order to replace consolePort with consoleSocket.

I'm not sure which path I should put. Other options are:
- simply remove the "consolePort" entries and rely on the default value of IPCM
- use abstract sockets (e.g. "@ipcm-console"), which looks acceptable for testing purpose

> Could you post an example on how to use socat to connect to the socket file?

    socat - UNIX:/run/ipcm-console.sock

If you have socat built with readline:

    socat READLINE UNIX:/run/ipcm-console.sock

With abstract sockets:

    socat - ABSTRACT:ipcm-console

**Comment by vmaffione** [comment]
2016-03-31 21:35 GMT+02:00 Julien Muchembled <notifications@github.com>:

> About runstatedir, localstatedir does not work for a system setup, because
> it would produce /var/ipcm-console.sock
> So I looked at why runstatedir was empty for you and found that it's a
> recent feature (Autoconf 2.70 is not even released yet; it was backported
> by Debian in version 2.69-9).
> I'd rather add the following lines just before the sed command:
>
> # --runstatedir is new to Autoconf 2.70
> runstatedir ?= $(localstatedir)/run
>
> Ok, please go on, then I'll give a test to check whether it works.


> Also, an additional commit is required to change all the sample
> configuration files stored in tests/conf, in order to replace consolePort
> with consoleSocket.
>
> I'm not sure which path I should put. Other options are:
>
>    - simply remove the "consolePort" entries and rely on the default
>    value of IPCM
>    - use abstract sockets (e.g. "@ipcm-console"), which looks acceptable
>    for testing purpose
>
> The examples assume --prefix /usr/local/irati, so I would just use
````
"consoleSocket": "/usr/local/irati/var/run/ipcm-console.sock",
````


>
>
> Could you post an example on how to use socat to connect to the socket
> file?
>
> socat - UNIX:/run/ipcm-console.sock
>
> If you have socat built with readline:
>
> socat READLINE UNIX:/run/ipcm-console.sock
>
> With abstract sockets:
>
> socat - ABSTRACT:ipcm-console
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/IRATI/stack/pull/883#issuecomment-204092920>
>



-- 
Vincenzo Maffione

**Comment by vmaffione** [comment]
There is still a problem due to the fact that var/run may not already exist.
I should have fixed it in this commit

https://github.com/vmaffione/stack/commit/20b58bcc7d3d4f62e952474389da020738b90899

for some reason I'm not able to create a pull request against your fork, it seems github is buggy tonight.

@jmuchemb , can you git-am/cherry-pick this commit on top of this pull request? Or provide an alternative fix?

**Comment by jmuchemb** [comment]
Sorry, I should have checked with custom --prefix
I amended your commit to use `$(localstatedir)/run` instead of `$(prefix)/var/run`

**Comment by jmuchemb** [comment]
I amended again to fully support --runstatedir. In order to fix `DEFAULT_RUNDIR` without repeating `runstatedir ?= $(localstatedir)/run` 3 times, I also moved the `install-data-local:` target to `rinad/src/ipcm/Makefile.am`

**Comment by vmaffione** [comment]
This looks good! Merging

**Pull request merged**